impeller/toolkit/android/toolkit_android_unittests.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/synchronization/waitable_event.h"
+#include "flutter/testing/testing.h"
+#include "impeller/toolkit/android/hardware_buffer.h"
+#include "impeller/toolkit/android/proc_table.h"
+#include "impeller/toolkit/android/surface_control.h"
+#include "impeller/toolkit/android/surface_transaction.h"
+
+namespace impeller::android::testing {
+
+class ToolkitAndroidTest : public ::testing::Test {
```

Comment:
I don't know if you want to do this as a follow up, but related to Jonah's question:

The tests here don't have as much coverage as the ndk_helpers_unittests (specifcially, there aren't assertions that the API works and that we can allocate/destroy these objects or that the API symbols are correctly resolved). Maybe this will be good for a follow up patch, but consider porting those tests to this toolkit, deleting the FML versions, and porting the Android embedding to use this toolkit?

If that's not possible, I'd rather not land this in Impeller - I'm somewhat concerned about having more places that we do these lookups and introduce typos/untested code, especially since we just got that cleaned up.

Reply:
> If that's not possible, I'd rather not land this in Impeller - I'm somewhat concerned about having more places that we do these lookups and introduce typos/untested code, especially since we just got that cleaned up.

+1

Reply:
> API works and that we can allocate/destroy these objects.

`CanCreateHardwareBuffer` for hardware buffers. `CanApplySurfaceTransaction` for surface transactions. Same as the NDK helpers.

> API symbols are correctly resolved.

`CanCreateProcTable`. You'll only get a valid proc table if the API symbols are correctly resolved and the version checks pass. For the individual subcomponents, this is also checked with `IsAvailableOnPlatform`.

> The tests here don't have as much coverage.

I'm surprised by this statement. There are only two tests in NDK helpers related to anything in this patch. `HardwareBuffer` and `SurfaceTransaction`. The entire corpus is only 5 tests in total ([unless I am missing something](https://github.com/flutter/engine/blob/main/fml/platform/android/ndk_helpers_unittests.cc#L79C1-L80C1)). The two tests I mentioned above more than cover the the assertions in NDK helpers.

I want to use the toolkit for everything and remove the NDK helpers after porting all existing tests. But, at least till I get to the choreographer, there isn't going to be anything to port as it should already be covered.

Reply:
We seem to be missing coverage here of `AHardwareBuffer_describe`, `AHardwareBuffer_isSupported`, and `AHardwareBuffer_getId`, which are needed by the current Android implementation.

Reply:
Yeah, but the toolkit has no ability to describe or get the ID of those objects yet. Once I work on adding those features get rid of NDK helpers, I can patch those in.

Reply:
The AHB code in ndk_helpers is used somewhere currently right? Even if not all of it.

Where is there at est in this file that creates a buffer, verfies its descriptor matches what it created, and verifies the buffer can be destroyed?

Reply:
> Yeah, but the toolkit has no ability to describe or get the ID of those objects yet. Once I work on adding those features get rid of NDK helpers, I can patch those in.

I would rather see either Impeller use the FML utilities until it's ready to replace them with something better, or the rest of the code base using new improved utilities with similar test coverage. Otherwise we're likely to end up in a state where we have multiple locations to do NDK related lookups with different available/missing test coverage ...

Reply:
All I want to do is to facilitate landing https://github.com/flutter/engine/pull/51213. The ask on that patch was that it was too large and it should be split up. I have split it up into https://github.com/flutter/engine/pull/51295, https://github.com/flutter/engine/pull/51298, and this one. Two more patches are incoming after this to facilitate the request for smaller, more self-contained PRs.

I am on the hook for killing the FML utilities. I can kill the FML utilities in this patch if this gets us past the bikeshedding but it will make this patch bigger. Or, kill the FML utilities immediately after this lands if that seems more elegant.

Reply:
Ping @dnfield. Do you have a preference? Want to unblock this review ASAP.

Reply:
I would rather eithe rkill it in this patch or update the AHB swapchain patch to use the FML utilities.